### ISSUE TITLE
three outputs

### ISSUE THREAD

**Issue by shockjiang** [halide/Halide#2912]
I am trying to return 3 outputs in a halide pipeline. When I turned the code to *.h and windows obj files, the function signature is like this:
int matching_x86_64_windows(struct halide_buffer_t *_input_buffer,  struct halide_buffer_t *_aggregate_buffer) HALIDE_FUNCTION_ATTRS;

I created a halide_buffer t[3] as the output and pass it to the function:
halide_buffer_t input;
halide_buffer_t out[3];
matching_x86_64_windows(&input, &out[0]).

But it seems like this is wrong way of calling the function. And I cannot find any document for this. Could you help?

**Comment by MDBrothers** [comment]
Did you define the output function as a tuple? (http://halide-lang.org/tutorials/tutorial_lesson_13_tuples.html)

**Comment by shockjiang** [comment]
yes, I did. In fact, when there are two outputs, it works. But when there are three, it behaves like what I described.

**Comment by shockjiang** [comment]
The example shows two outputs, but I have three. I used python bindings, see the code below:

```
        ...
        re[x,y] = [e, e2, e3]
        
        target = hl.Target()
        target.os = hl.TargetOS.Windows
        target.arch = hl.TargetArch.X86
        target.bits = 64
        target.set_features([hl.TargetFeature.AVX, hl.TargetFeature.SSE41])
        args = [lin, rin, llow, lupper, rlow, rupper, block_x, block_y, min_disp, max_disp, noise_epsilon]
        key = "matching"
        re.compile_to_file(key+"_x86_64_windows", args, key+"_x86_64_windows", target)   
```

**Comment by MDBrothers** [comment]
I'm afraid my understanding is exhausted. (I'm not a Halide dev, just an over-eager user).  I just have a few thoughts that could maybe help or help you sidestep the problem.

I'm not a master Python programmer, but the syntax you used:
` re[x,y] = [e, e2, e3]`
Looks like you're setting the contents of re[x,y] to a list rather than a tuple of expressions, is that right? Could this be failing in a confusing way?

Other than that, one way to get around this might b to have re as a 3 or n channel image, and as long as you can use the same data type for each, just copy your expressions to the channels of your output.

**Comment by steven-johnson** [comment]
Technically, you're assigning a List and not a Tuple to re; it should be:

        re[x,y] = (e, e2, e3)


That said, this really should fail loudly rather than generate a misleading
result.

**Comment by matthewwen** [comment]
Was there a solution to this? I too want have 3 buffers, and I want to combine it to make 1 image. Each buffer representing an image.

**Comment by abadams** [comment]
**Automated triage note:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — that is the expected outcome for some fraction of these, and no offence taken.

**Reason (CLOSE_FIXED):** The reporter assigned a Python list rather than a tuple to a Func, which in the old bindings produced a misleading single-output result; the thread's conclusion was that this should not silently misbehave. In the current pybind11 bindings a list of Exprs converts to a Halide::Tuple explicitly, so `f[x, y] = [e1, e2, e3]` yields a genuine 3-element tuple (and hence three output buffers in the AOT signature), and anything non-convertible raises a value_error.

**Evidence:** python_bindings/src/halide/halide_/PyTuple.cpp: `py::implicitly_convertible<std::vector<Expr>, Tuple>()` and `to_halide_tuple()` which throws `py::value_error("Expected an Expr or tuple-of-Expr.")` on bad input.

Comment and close: "The current Python bindings convert a list-of-Expr to a Halide Tuple explicitly (see PyTuple.cpp), so `f[x, y] = [e1, e2, e3]` is a real 3-tuple and AOT compilation emits three output halide_buffer_t arguments; malformed right-hand sides now raise instead of silently misbehaving. Closing as fixed — please reopen with a current repro if you still see a single aggregated output."
